code_map.py

```diff
@@ -138,23 +143,103 @@ def get_code_map_view():
 
-def refresh_map_for(view):
-    file = view.file_name()
-    if file and code_map_generator.can_map(file):
-        map_view = get_code_map_view()
-        if map_view:
-            map_view.run_command('code_map_generator', {"source": file})
+    for i, col in enumerate(cols):
+        if col > 0:
+            cols[i] = col*width
 
-            # avoid certain error messages
-            try:
-                if map_view.sel():
-                    lines = map_view.split_by_newlines(
-                        map_view.sel()[0])
-                if len(lines) > 1:
-                    map_view.sel().clear()
-            except:
+    cols.append(1)
+    newcell = [last_col, 0, last_col + 1, last_row]
+    cells.append(newcell)
+    groups = w.num_groups()
+    w.run_command("set_layout", layout)
+    sublime.set_timeout(lambda: Mapper.block_max_pane(False), 10)
+    return (groups + 1)
+
+# -----------------
+
+def reset_layout(reduce):
+    '''Removes the Code Map group, and scales up the layout'''
+    w = win()
+    layout = w.get_layout()
+    cols = layout['cols']
+    width = 1 - settings().get("codemap_width")
+
+    map_view = get_code_map_view()
+    if map_view:
+        w.set_view_index(map_view, 0, 0)
+
+    if reduce:
+        for i, col in enumerate(cols):
+            if col > 0:
+                cols[i] = col/width
+
+        cols[-2] = 1.0
+        del cols[-1]
+        del layout['cells'][-1]
+
+    Mapper.block_max_pane(True)
+    w.run_command("set_layout", layout)
+    sublime.set_timeout(lambda: Mapper.block_max_pane(False), 10)
+
+# -----------------
+
+def refresh_map_for(view, from_view=False):
+    global CURRENT_TEMP_ID
+
+    # -----------------
+
+    def clear_map_selection():
+        # avoid certain error messages and clear abnormal selections
+        try:
+            if map_view.sel():
+                lines = map_view.split_by_newlines(
+                    map_view.sel()[0])
+            if len(lines) > 1:
                 map_view.sel().clear()
+        except:
+            map_view.sel().clear()
+
+    # -----------------
+
+    def generate_from(file):
+        global Generated_Map
+
+        map = code_map_generator.get_mapper(file)
+        if map:
+            Generated_Map = map
+            map_view.run_command('code_map_generator', {"source": file})
+            clear_map_selection()
+
+    # -----------------
+
+    file = view.file_name()
+    map_view = get_code_map_view()
+    CURRENT_TEMP_ID = None
+
+    if not map_view:
+        return
+    elif file and path.basename(file) == "Code - Map":
+        return
+    elif file:
+        generate_from(file)
+    elif from_view or view.id() in TEMP_VIDS:
+        if view.id() not in TEMP_VIDS:
+            TEMP_VIDS.append(view.id())
+        CURRENT_TEMP_ID = view.id()
+        map_view.run_command('code_map_generator', {"source": view.id()})
+        clear_map_selection()
+    elif not file:
+        win().run_command('code_map_temp_view_msg')
 
```

Comment:
What has been done here:

- if it's a file, instead of running can_map(file), all checks are made here, and the map is being created already, stored in Generated_Map, then code_map_generator is run.
- if it's a view instead, the code_map_generator is run with a non-str argument(the view id is int), so that it knows we're creating the map from a view and not from a file. The views info are stored, so that the view and the map will be remembered.